File: 2022/src/day02.scala
import scala.io.Source.fromFile
import scala.util.Using

val alias = Map("X" -> "A", "Y" -> "B", "Z" -> "C")
val weakness = Map("A" -> "B", "B" -> "C", "C" -> "A")
val points = Map("A" -> 1, "B" -> 2, "C" -> 3)

def main(args: Array[String]): Unit = {
    val input = Using(fromFile("input/day02"))(_.mkString).get
    val rounds = input.split("\n").map(_.split(" "))
    val p1 = rounds.map({ case Array(op, me) => scoreMe(alias(me), op) }).sum
    val p2 = rounds.map({ case Array(op, strat) => scoreMe(move(op, strat), op) }).sum
    println(s"Part1: $p1")
    println(s"Part2: $p2")
}

def scoreMe(me: String, op: String): Int =
    points(me) + (if (op == me) 3 else if (weakness(me) != op) 6 else 0)

def move(op: String, strategy: String): String =
    strategy match {
        case "X" => weakness.find(_._2 == op).map(_._1).get
        case "Y" => op
        case _   => weakness(op)
    }

File: 2023/src/day02.ex
defmodule Day02 do
  @alias %{
    "X" => "A",
    "Y" => "B",
    "Z" => "C"
  }

  @weakness %{
    "A" => "B",
    "B" => "C",
    "C" => "A"
  }

  @points %{
    "A" => 1,
    "B" => 2,
    "C" => 3
  }

  def main do
    rounds = get_rounds()
    p1 = Enum.map(rounds, fn [op, me] -> score(@alias[me], op) end) |> Enum.sum()
    p2 = Enum.map(rounds, fn [op, strat] -> score(move(op, strat), op) end) |> Enum.sum()
    IO.puts("Part1: #{p1}")
    IO.puts("Part2: #{p2}")
  end

  def get_rounds() do
    File.read!("input/day02")
    |> String.split("\n")
    |> Enum.map(fn r -> String.split(r, " ") end)
  end

  def score(me, op) do
    @points[me] +
      cond do
        op == me -> 3
        @weakness[me] != op -> 6
        true -> 0
      end
  end

  def move(op, strat) do
    case strat do
      "X" -> Enum.find(@weakness, fn {_, v} -> v == op end) |> elem(0)
      "Y" -> op
      _ -> @weakness[op]
    end
  end
end

Day02.main()
